docs/advanced_topics.rst

```diff
@@ -0,0 +1,43 @@
+When ``DALI_AFFINITY_MASK`` is empty affinity is set based on the result returned by nvmlDeviceGetCpuAffinity.
+If there are more threads than values in ``DALI_AFFINITY_MASK`` value from nvmlDeviceGetCpuAffinity will be used for a thread without value provided by an environment variable,
+I.e.:
+
+.. code-block:: bash
+
+  num_threads=5
+  DALI_AFFINITY_MASK=3,5,6,10
+
+Will set thread 0 to CPU 3, 1 to CPU 5, 2 to CPU 6, 3 to CPU 10 and 4 to CPU which is returned by nvmlDeviceGetCpuAffinity.
+
+Memory consumption
+------------------
+
+DALI architecture makes all GPU operators processing the whole batch at the time and CPU one sample at the time. As DALI is only enlarging its memory, never shrinking, each GPU buffer is as large as the largest possible batch, while CPU buffers are as large as batch size multiplied by the size of the largest sample (even though CPU processes one sample at the time per thread still vector of samples need to reside in the memory). This lazy allocation strategy helps in cases when memory allocation is expensive like pinned CPU memory or GPU memory reducing the number of total memory operations. This is most visible for the operators which output size may differ from sample to sample, from run to run. Operator with the fixed size outputs - like crop is not affecting overall memory consumption growth over time.
+
+Operator buffer presizing
+-------------------------
+
+The rationale behind this functionality is allowing the user to fine tune his processing pipeline when he can precisely forecast memory consumption during DALI run saving the overhead of some reallocations.
+DALI uses intermediate buffers to pass data between operators in the processing graph. Due to internal DALI architecture, memory is newer freed but just enlarged when present buffers are not able to house data they need to fit in. However, in some cases, even this limited number of allocations still could affect DALI performance. If the user knows how much memory each operator buffer may need, it is possible to provide a hint to presize buffers before the first run.
+Two parameters are available - ``bytes_per_sample`` pipeline argument accepting one value, that is used globally across all operators for all buffers. And ``bytes_per_sample_hint`` per operator argument which accepts one value or list of values. When one value is provided it is used for all output buffers for a given operator. When the list is provided each buffer is presized to the corresponding size.
+
```

Comment:
Two parameters are available: First, the ``bytes_per_sample`` pipeline argument, which accepts one value that is used globally across all operators for all buffers. 

The second parameter is the ``bytes_per_sample_hint`` per operator argument, which accepts one value or a list of values. When one value is provided it is used for all output buffers for a given operator. When a list is provided then each buffer is presized to the corresponding size.

Reply:
Adjusted